src/rational/q/from.rs

```diff
@@ -213,15 +216,29 @@ impl Q {
     /// let a: Q = Q::from_f64(-123.4567);
     /// ```
     pub fn from_f64(value: f64) -> Self {
-        let f = Fraction::from(value);
-        let sign = f
-            .sign()
-            .expect("Got None element instead of a fraction, may be overflow error (NaN)")
-            .is_positive();
-        match sign {
-            true => Q::try_from((f.numer().unwrap(), f.denom().unwrap())).unwrap(),
-            false => Q::try_from((f.numer().unwrap(), f.denom().unwrap())).unwrap() * Q::MINUS_ONE,
-        }
+        let bits: u64 = value.to_bits();
+        let sign = if bits >> 63 == 0 {
+            Q::ONE
+        } else {
+            Q::MINUS_ONE
+        };
+        let mut exponent: i16 = ((bits >> 52) & 0x7ff) as i16;
+        let mantissa = if exponent == 0 {
+            (bits & 0xfffffffffffff) << 1
+        } else {
+            // prepend one bit to the mantissa because the most significant bit is implicit
+            (bits & 0xfffffffffffff) | 0x10000000000000
+        };
+
+        // -1023 because of the offset representation of the exponent
+        // -52 because the mantissa is 52 bit long
+        exponent -= 1023 + 52;
+        let shift = match exponent {
+            e if e >= 1 => Q::from(2).pow(e).unwrap(),
+            e => Q::try_from((&1, &2)).unwrap().pow(e.abs()).unwrap(),
```

Comment:
```suggestion
            // This could be optimized with `fmpz_lshift_mpn` once it is part of flint_sys.
            e if e >= 1 => Q::from(2).pow(e).unwrap(),
            e => Q::try_from((&1, &2)).unwrap().pow(e.abs()).unwrap(),
```
This is not super efficient, but just add a comment for now.